scripts/update_xspec_functions.py

```diff
@@ -36,68 +36,450 @@
 
   // End model definitions
 
-marks but please note that you will need to add back in the
+marks. To try and support old versions the script takes the current
+_xspec.cc file and uses that as a basis for this section, so that
+existing version checks will remain, but this is only meant as a
+starting point, so check the output.
 
-    #ifdef XSPEC_x_y_x
-    #else
-    #endif
-
-lines as there's no way to identify them (i.e. add them automatically)
-from just a single model.dat file.
+The XSPEC version should be given in dotted form - e.g. 12.15.0 - and
+without any patch level.
 
 """
 
-from typing import Optional
+from importlib import resources
+from pathlib import Path
 import sys
+from typing import Literal
 
 from sherpa.astro.utils import xspec
 
 
-def report_model_definitions(models: list[xspec.ModelDefinition]) -> None:
-    """Print to stdout the code needed to bind to the models."""
+# Should really provide better typing
+ModelName = str
+ModelDefinition = str
 
-    prev = None
-    for model in models:
-        prev = print_model(model, prev)
+# A version can be major, minor, micro or, to support '#else' lines,
+# the string "older", but that's only for the "state" handling.
+#
+Version = tuple[int, int, int]
+StateVersion = Version | Literal["older"]
 
+State = tuple[ModelName, ModelDefinition, StateVersion | None] | None
 
-def print_model(model: xspec.ModelDefinition,
-                prev: Optional[str]) -> Optional[str]:
-    """Print out a model"""
+# Store a set of #ifdef...#else lines by
+# - a dictionary where the key is each "#" line and the
+#   values are the lines guarded by that check
+# - a checksum value which just gives a combination of
+#   all the "#.." lines so we can easily look for
+#   a compatible set of lines.
+#
+Checksum = str
+Storage = tuple[dict[str, list[str]], Checksum]
+
+NO_CHECKSUM: Checksum = "<none>"
+
+START_MACRO = "// Start model definitions"
+END_MACRO = "// End model definitions"
 
-    try:
-        code = xspec.model_to_compiled(model)
-    except ValueError:
-        # Assume an unsupported model
-        # sys.stderr.write(f"# Unsupported: {model.name} / {model.funcname}\n")
-        return prev
 
-    # For the moment models that are calculated per-spectrum are unsupported.
+def convert_version(smajor: str, sminor: str, smicro: str) -> Version:
+    """Convert version tuple"""
+
```

Comment:
Isn't this identical to the function in `add_xspec_model.py`? If so, write once and import. But maybe that's changed in a latter commit; hard to know when reviewing commit-by-commit. Might have other get back to this comment again in the end.

Reply:
So, I did think about moving some code into a "helper" module that these scripts could import, but I decided that I didn't want to spend the extra time to do it.